File: mytradebot/api/serializers.py
"""文件用途：把内部数据结构序列化成 WebSocket / REST 响应 payload。"""
from __future__ import annotations

import json
from datetime import datetime, timezone
from typing import Any

from ..config import AppConfig
from ..domain.price_action import Candle
from ..domain.quotes import QuoteState
from ..market_data.router import MarketInstrument

THUMBNAIL_CANDLE_LIMIT = 60
DEFAULT_AGENT_USER_PROMPT = "Analyze the current K-line chart and update the watch plan."


def utc_now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


def agent_session_title(instrument: MarketInstrument) -> str:
    return f"{instrument.label} · {instrument.symbol}"


def agent_session_config_kwargs(config: Any) -> dict[str, Any]:
    return {
        "api_mode": config.api_mode,
        "reasoning_effort": config.reasoning_effort,
    }


def sse_event(event: dict[str, Any]) -> str:
    return f"data: {json.dumps(event, ensure_ascii=False, separators=(',', ':'))}\n\n"


def instrument_payload(instrument: MarketInstrument, *, default_interval: str) -> dict[str, Any]:
    return {
        "key": instrument.key,
        "symbol": instrument.symbol,
        "label": instrument.label,
        "source": instrument.source,
        "instType": getattr(instrument, "inst_type", None),
        "group": instrument.group,
        "analysisInterval": instrument.analysis_interval or default_interval,
    }


def instrument_catalog_item_payload(
    instrument: MarketInstrument,
    *,
    active_keys: set[str],
) -> dict[str, Any]:
    base_asset = getattr(instrument, "base_asset", instrument.symbol)
    quote_asset = getattr(instrument, "quote_asset", "")
    inst_type = getattr(instrument, "inst_type", None)
    group = getattr(instrument, "group", None)
    category = getattr(instrument, "category", None)
    dex = getattr(instrument, "dex", None)
    if instrument.source == "bitget":
        display_text = f"{inst_type} · {base_asset}/{quote_asset}"
    else:
        prefix = (category or "perp").upper()
        suffix = f" · {dex}" if dex else ""
        display_text = f"{prefix} perp · {base_asset}/{quote_asset}{suffix}"
    return {
        "source": instrument.source,
        "symbol": instrument.symbol,
        "label": instrument.label,
        "instType": inst_type,
        "group": group,
        "category": category,
        "dex": dex,
        "key": instrument.key,
        "displayText": display_text,
        "exists": instrument.key in active_keys,
    }


def candle_payload(candle: Candle) -> dict[str, Any]:
    return {
        "time": candle.open_time_ms // 1000,
        "open": candle.open,
        "high": candle.high,
        "low": candle.low,
        "close": candle.close,
        "volume": candle.volume,
    }


def quote_payload(
    quote: QuoteState,
    *,
    stale_after_seconds: int,
) -> dict[str, Any]:
    return {
        "symbol": quote.symbol,
        "displayName": quote.display_name,
        "price": quote.price,
        "priceLabel": quote.price_label(),
        "change": quote.change,
        "changePercent": quote.change_percent,
        "changeLabel": quote.change_label(),
        "percentLabel": quote.percent_label(),
        "previousClose": quote.previous_close,
        "dayHigh": quote.day_high,
        "dayLow": quote.day_low,
        "volume": quote.volume,
        "volumeLabel": quote.volume_label(),
        "currency": quote.currency,
        "exchange": quote.exchange,
        "status": quote.status,
        "ageLabel": quote.age_label(),
        "stale": quote.is_stale(stale_after_seconds),
        "lastError": quote.last_error,
        "updateCount": quote.update_count,
        "multiTimeframeIntervals": sorted(quote.multi_timeframe_candles.keys()),
        "candles": [candle_payload(candle) for candle in quote.candles],
        "thumbnailCandles": [
            candle_payload(candle)
            for candle in quote.thumbnail_candles[-THUMBNAIL_CANDLE_LIMIT:]
        ],
    }


def serialize_market_state(
    *,
    config: AppConfig,
    instruments: tuple[MarketInstrument, ...],
    quotes: dict[str, QuoteState],
    stream_status: str,
    agent_analyses: dict[str, dict[str, Any]] | None = None,
    open_trades: list[dict[str, Any]] | None = None,
    exchange_positions: list[dict[str, Any]] | None = None,
    exchange_orders: list[dict[str, Any]] | None = None,
    recent_news: list[dict[str, Any]] | None = None,
    news_status: dict[str, Any] | None = None,
) -> dict[str, Any]:
    groups: dict[str, list[str]] = {}
    for instrument in instruments:
        groups.setdefault(instrument.source, []).append(instrument.key)
    return {
        "type": "state",
        "updatedAt": utc_now_iso(),
        "streamStatus": stream_status,
        "config": {
            "analysis": {
                "enabled": config.analysis.enabled,
                "interval": config.analysis.interval,
                "lookback": config.analysis.lookback,
                "pollIntervalSeconds": config.analysis.poll_interval_seconds,
                "staleAfterSeconds": config.analysis.stale_after_seconds,
            },
            "agent": {
                "enabled": config.agent.enabled,
                "provider": config.agent.provider,
                "apiMode": config.agent.api_mode,
                "model": config.agent.model,
                "maxCandles": config.agent.max_candles,
                "reasoningEffort": config.agent.reasoning_effort,
                "providerProfiles": {
                    name: {
                        "enabled": profile.enabled,
                        "models": list(profile.models),
                        "modelEfforts": dict(profile.model_efforts),
                    }
                    for name, profile in config.agent.provider_profiles.items()
                },
            },
            "display": {
                "refreshIntervalMs": config.display.refresh_interval_ms,
                "staleAfterSeconds": config.display.stale_after_seconds,
                "stockPollIntervalSeconds": config.display.stock_poll_interval_seconds,
            },
            "news": {
                "enabled": config.news.enabled,
                "pollIntervalSeconds": config.news.poll_interval_seconds,
                "maxIntervalSeconds": config.news.max_interval_seconds,
                "recentLimit": config.news.recent_limit,
                "reutersUrl": config.news.reuters_url,
                "requestTimeoutSeconds": config.news.request_timeout_seconds,
                "retentionDays": config.news.retention_days,
            },
            "socialFeed": {
                "enabled": config.social_feed.enabled,
                "recentLimit": config.social_feed.recent_limit,
                "retentionDays": config.social_feed.retention_days,
                "maxItems": config.social_feed.max_items,
            },
            "memory": {
                "enabled": config.memory.enabled,
                "useMemories": config.memory.use_memories,
                "generateMemories": config.memory.generate_memories,
                "extractModel": config.memory.extract_model,
                "consolidationModel": config.memory.consolidation_model,
                "maxRawMemories": config.memory.max_raw_memories_for_consolidation,
                "maxUnusedDays": config.memory.max_unused_days,
                "maxSourceAgeDays": config.memory.max_source_age_days,
                "maxRolloutsPerStartup": config.memory.max_rollouts_per_startup,
                "minSessionIdleHours": config.memory.min_session_idle_hours,
                "extensionRetentionDays": config.memory.extension_retention_days,
            },
            "trading": {
                "hyperliquidEnabled": config.trading.hyperliquid_enabled,
                "bitgetDemoEnabled": config.trading.bitget_demo_enabled,
            },
            "sourcePath": str(config.source_path) if config.source_path else None,
        },
        "instruments": [
            instrument_payload(instrument, default_interval=config.analysis.interval)
            for instrument in instruments
        ],
        "groups": groups,
        "quotes": {
            key: quote_payload(
                quote,
                stale_after_seconds=config.display.stale_after_seconds,
            )
            for key, quote in quotes.items()
        },
        "agentAnalyses": agent_analyses or {},
        "openTrades": open_trades or [],
        "exchangePositions": exchange_positions or [],
        "exchangeOrders": exchange_orders or [],
        "recentNews": recent_news or [],
        "newsStatus": news_status or {},
    }

File: tradex/api/serializers.ts
import { AppConfig } from "../config/index.js";
import { QuoteState } from "../domain/quotes.js";
import { MarketInstrument } from "../market_data/router.js";
import { ExchangeOrder, ExchangePosition, orderToPayload, positionToPayload } from "../trading/exchange_models.js";
import { Trade, tradeToPayload } from "../trading/models.js";
import { NewsItem, newsItemToPayload } from "../news/types.js";
import type { Jin10CalendarEvent, Jin10Quote, Jin10Status } from "../jin10/types.js";

export function serializeState(input: {
  config: AppConfig;
  instruments: readonly MarketInstrument[];
  quotes: Record<string, QuoteState>;
  streamStatus: string;
  openTrades: Trade[];
  exchangePositions: ExchangePosition[];
  exchangeOrders: ExchangeOrder[];
  recentNews: NewsItem[];
  newsStatus?: Record<string, unknown>;
  jin10?: {
    status: Jin10Status;
    calendar: Jin10CalendarEvent[];
    quotes: Jin10Quote[];
  };
  options?: {
    snapshots: Record<string, unknown>;
  } | null;
}): Record<string, unknown> {
  const groups: Record<string, string[]> = {};
  for (const instrument of input.instruments) {
    const group = instrument.source || "other";
    groups[group] = [...(groups[group] ?? []), instrument.key];
  }

  // Determine whether instruments from a given source support agent analysis
  // (candles, multi-timeframe). Sources that only provide quote snapshots are not analysable.
  const sourceAnalysable = resolveSourceAnalysable(input.config);

  // Build base instruments and quotes
  const instruments = input.instruments.map((instrument) => ({
    key: instrument.key,
    symbol: instrument.symbol,
    label: instrument.label,
    source: instrument.source,
    instType: "instType" in instrument ? instrument.instType : null,
    group: instrument.group,
    analysisInterval: instrument.analysisInterval || input.config.analysis.interval,
    analysable: sourceAnalysable(instrument.source),
  }));
  const quotes: Record<string, unknown> = Object.fromEntries(
    Object.entries(input.quotes).map(([key, quote]) => [key, serializeQuote(quote, input.config.display.staleAfterSeconds)]),
  );

  // Inject Jin10 quotes into watchlist when available
  if (input.jin10 && input.jin10.quotes.length > 0) {
    const jin10Keys: string[] = [];
    for (const q of input.jin10.quotes) {
      const key = `jin10:${q.code}`;
      jin10Keys.push(key);
      instruments.push({
        key,
        symbol: q.code,
        label: q.name || q.code,
        source: "jin10",
        instType: null,
        group: "jin10",
        analysisInterval: "",
        analysable: sourceAnalysable("jin10"),
      });
      quotes[key] = serializeJin10Quote(q);
    }
    groups["jin10"] = jin10Keys;
  }

  return {
    type: "state",
    updatedAt: new Date().toISOString(),
    streamStatus: input.streamStatus,
    config: serializeConfig(input.config),
    instruments,
    groups,
    quotes,
    agentAnalyses: {},
    openTrades: input.openTrades.map((trade) => tradeToPayload(trade)),
    exchangePositions: input.exchangePositions.map(positionToPayload),
    exchangeOrders: input.exchangeOrders.map(orderToPayload),
    recentNews: input.recentNews.map(newsItemToPayload),
    newsStatus: input.newsStatus ?? { enabled: input.config.news.enabled },
    jin10: input.jin10 ?? null,
    options: input.options ?? null,
  };
}

export function serializeConfig(config: AppConfig): Record<string, unknown> {
  return {
    analysis: config.analysis,
    agent: {
      enabled: config.agent.enabled,
      provider: config.agent.provider,
      apiMode: config.agent.apiMode,
      model: config.agent.model,
      maxCandles: config.agent.maxCandles,
      candleContextMode: config.agent.candleContextMode,
      reasoningEffort: config.agent.reasoningEffort,
      providerProfiles: Object.fromEntries(
        Object.entries(config.agent.providerProfiles).map(([name, profile]) => [
          name,
          {
            enabled: profile.enabled,
            models: profile.models,
            modelEfforts: Object.fromEntries(profile.modelEfforts),
            baseUrl: profile.baseUrl || undefined,
            apiKeyConfigured: Boolean(profile.apiKey),
            apiKeyFromEnv: providerApiKeyFromEnv(name),
            customModels: profile.customModels,
          },
        ]),
      ),
    },
    display: config.display,
    news: config.news,
    memory: {
      enabled: config.memory.enabled,
      useMemories: config.memory.useMemories,
      generateMemories: config.memory.generateMemories,
      disableOnExternalContext: config.memory.disableOnExternalContext,
      storagePath: config.memory.storagePath,
      extractModel: config.memory.extractModel,
      consolidationModel: config.memory.consolidationModel,
      maxRawMemories: config.memory.maxRawMemoriesForConsolidation,
      maxUnusedDays: config.memory.maxUnusedDays,
      maxSourceAgeDays: config.memory.maxSourceAgeDays,
      maxRolloutsPerStartup: config.memory.maxRolloutsPerStartup,
      minSessionIdleHours: config.memory.minSessionIdleHours,
      extensionRetentionDays: config.memory.extensionRetentionDays,
    },
    socialFeed: config.socialFeed,
    trading: config.trading,
    mcp: {
      enabled: config.mcp.enabled,
      configPath: config.mcp.configPath,
    },
    jin10: {
      enabled: config.jin10.enabled,
      tokenConfigured: Boolean(config.jin10.token),
      flashEnabled: config.jin10.flashEnabled,
      calendarEnabled: config.jin10.calendarEnabled,
      quotesEnabled: config.jin10.quotesEnabled,
      quotesCodes: config.jin10.quotesCodes,
    },
    options: {
      enabled: config.options.enabled,
      provider: config.options.provider,
      symbols: config.options.symbols,
      pollIntervalSeconds: config.options.pollIntervalSeconds,
      strikeRangePercent: config.options.strikeRangePercent,
      tradier: config.options.tradier ? {
        apiKeyConfigured: Boolean(config.options.tradier.apiKey),
        baseUrl: config.options.tradier.baseUrl,
      } : undefined,
      deribit: config.options.deribit,
    },
    sourcePath: config.sourcePath,
  };
}

function providerApiKeyFromEnv(provider: string): boolean {
  if (provider === "anthropic") return Boolean(process.env.ANTHROPIC_API_KEY || process.env.ANTHROPIC_AUTH_TOKEN);
  return false;
}

function serializeQuote(quote: QuoteState, staleAfterSeconds: number): Record<string, unknown> {
  return {
    symbol: quote.symbol,
    displayName: quote.displayName,
    price: quote.price,
    priceLabel: quote.priceLabel(),
    change: quote.change,
    changePercent: quote.changePercent,
    changeLabel: quote.changeLabel(),
    percentLabel: quote.percentLabel(),
    previousClose: quote.previousClose,
    dayHigh: quote.dayHigh,
    dayLow: quote.dayLow,
    volume: quote.volume,
    volumeLabel: quote.volumeLabel(),
    currency: quote.currency,
    exchange: quote.exchange,
    status: quote.status,
    ageLabel: quote.ageLabel(),
    stale: quote.isStale(staleAfterSeconds),
    lastError: quote.lastError,
    updateCount: quote.updateCount,
  };
}

function serializeJin10Quote(q: Jin10Quote): Record<string, unknown> {
  const sign = q.change >= 0 ? "+" : "";
  const pctSign = q.changePercent >= 0 ? "+" : "";
  return {
    symbol: q.code,
    displayName: q.name || q.code,
    price: q.close,
    priceLabel: q.close ? formatJin10Price(q.close) : "-",
    change: q.change,
    changePercent: q.changePercent,
    changeLabel: `${sign}${q.change.toFixed(2)}`,
    percentLabel: `${pctSign}${q.changePercent.toFixed(2)}%`,
    previousClose: null,
    dayHigh: q.high,
    dayLow: q.low,
    volume: q.volume,
    volumeLabel: q.volume ? String(q.volume) : "-",
    currency: "",
    exchange: "jin10",
    status: q.time ? "active" : "idle",
    ageLabel: q.time || "-",
    stale: false,
    lastError: null,
    updateCount: 0,
  };
}

function formatJin10Price(price: number): string {
  if (price >= 100) return price.toFixed(2);
  if (price >= 1) return price.toFixed(4);
  return price.toFixed(5);
}

/**
 * Returns a predicate that determines whether instruments from a given source
 * support agent analysis (candles, multi-timeframe context).
 *
 * Policy:
 * - Sources with candle data (bitget, hyperliquid) default to true
 * - Quote-only sources (jin10) default to false
 * - Each source's config can override via `agent_analysis` field
 */
function resolveSourceAnalysable(config: AppConfig): (source: string) => boolean {
  // Build a source → analysable lookup from config.
  // Default is true for data sources that natively provide candles.
  const overrides: Record<string, boolean> = {
    jin10: config.jin10.agentAnalysis,
  };

  return (source: string): boolean => {
    if (source in overrides) return overrides[source];
    // Bitget, Hyperliquid, and any future candle-capable sources default to true
    return true;
  };
}
